mutation/mutation_partition_serializer.cc

```diff
@@ -182,6 +184,95 @@ void mutation_partition_serializer::write_serialized(Writer&& writer, const sche
     std::move(clustering_rows).end_rows().end_mutation_partition();
 }
 
+template<typename Output>
+class mutation_partition_serializer::write_serialized_executor {
+    const schema& _schema;
+    const mutation_partition& _mp;
+    ser::mutation_partition__static_row<Output> _static_row_writer;
+    struct range_tombstones_state {
+        ser::mutation_partition__range_tombstones<Output> writer;
+        range_tombstone_list::range_tombstones_type::const_iterator begin;
+        range_tombstone_list::range_tombstones_type::const_iterator end;
+    };
+    std::optional<range_tombstones_state> _range_tombstones;
+    struct clustering_rows_state {
+        ser::mutation_partition__rows<Output> writer;
+        mutation_partition::rows_type::const_iterator begin;
+        mutation_partition::rows_type::const_iterator end;
+    };
+    std::optional<clustering_rows_state> _clustering_rows;
+
+    std::function<stop_iteration(write_serialized_executor*)> _stage;
+
+public:
+    write_serialized_executor() = delete;
+
+    template<typename Writer>
+    explicit write_serialized_executor(Writer&& writer, const schema& s, const mutation_partition& mp)
+        : _schema(s)
+        , _mp(mp)
+        , _static_row_writer(std::move(writer).write_tomb(_mp.partition_tombstone()).start_static_row())
+        , _stage(std::mem_fn(&write_serialized_executor::write_static_row))
+    {}
+
+    stop_iteration run() {
+        return _stage(this);
+    }
+
+private:
+    stop_iteration write_static_row() {
+        _range_tombstones.emplace(write_row_cells(std::move(_static_row_writer), _mp.static_row().get(), _schema, column_kind::static_column)
+                .end_static_row()
+                .start_range_tombstones(),
+            _mp.row_tombstones().begin(), _mp.row_tombstones().end()
+        );
+        if (need_preempt()) {
+            _stage = std::mem_fn(&write_serialized_executor::write_range_tombstones);
+            return stop_iteration::no;
+        }
+        return write_range_tombstones();
+    }
+
+    stop_iteration write_range_tombstones() {
+        while (_range_tombstones->begin != _range_tombstones->end) {
+            const auto& rte = *_range_tombstones->begin;
+            auto& rt = rte.tombstone();
+            _range_tombstones->writer.add().write_start(rt.start).write_tomb(rt.tomb).write_start_kind(rt.start_kind)
+                .write_end(rt.end).write_end_kind(rt.end_kind).end_range_tombstone();
+            ++_range_tombstones->begin;
+            if (need_preempt()) {
```

Comment:
Instead of copy+pasting this serializer, can we instead just check add `if (thread::running_in_thread() && need_preempt())` to the existing serializer?

Reply:
`running_in_thread` is not absolutely reliable since it may return `true` in continuation-based async function or coroutines when they are called from a thread and before they yield (a.k.a schedule) the first time.
we can instead pass is_preemptible flag flag though, that would be based on the entry point.

Reply:
To eliminate any overhead of the extra logic maybe we can use a bool template parameter for IsPreemptible, but I'm not sure it's worth it.

Reply:
You know what, since we already have the `Output` template parameter we won't introduce a new template anyhow so I'll do it this way